File: Backend/src/Services/ProjectManagement/Softaxis.ProjectManagement.Domain/Entities/Sprint.cs
namespace Softaxis.ProjectManagement.Domain.Entities;

public sealed class Sprint
{
    private Sprint() { }

    public Sprint(Guid projectId, string name, string? goal, string? startDate, string? endDate, int sortOrder)
    {
        Id        = Guid.NewGuid();
        ProjectId = projectId;
        Name      = name.Trim();
        Goal      = goal?.Trim();
        StartDate = startDate;
        EndDate   = endDate;
        Status    = "planned";
        SortOrder = sortOrder;
        CreatedAt = DateTime.UtcNow;
    }

    public Guid     Id        { get; private set; }
    public Guid     ProjectId { get; private set; }
    public string   Name      { get; private set; } = string.Empty;
    public string?  Goal      { get; private set; }
    public string?  StartDate { get; private set; }
    public string?  EndDate   { get; private set; }
    public string   Status    { get; private set; } = "planned"; // planned | active | completed
    public int      SortOrder { get; private set; }
    public DateTime  CreatedAt { get; private set; }
    public DateTime? UpdatedAt { get; private set; }
    public bool      IsDeleted { get; private set; }

    public Project? Project { get; private set; }
    public ICollection<Issue> Issues { get; private set; } = new List<Issue>();

    public void UpdateDetails(string name, string? goal, string? startDate, string? endDate)
    {
        Name      = name.Trim();
        Goal      = goal?.Trim();
        StartDate = startDate;
        EndDate   = endDate;
        UpdatedAt = DateTime.UtcNow;
    }

    public void Start()
    {
        Status    = "active";
        UpdatedAt = DateTime.UtcNow;
    }

    public void Complete()
    {
        Status    = "completed";
        UpdatedAt = DateTime.UtcNow;
    }
}

File: BackendNode/src/modules/projectManagement/models/Sprint.js
const mongoose = require('mongoose');
const schema = new mongoose.Schema({
  tenantId:  { type: String, required: true, index: true },
  projectId: { type: String, required: true },
  name:      { type: String, required: true },
  goal:      String,
  startDate: String,
  endDate:   String,
  status:    { type: String, enum: ['planning', 'active', 'completed'], default: 'planning' },
  isDeleted: { type: Boolean, default: false },
  createdAt: { type: Date, default: Date.now },
  updatedAt: Date,
}, { versionKey: false });
schema.index({ tenantId: 1, projectId: 1 });
module.exports = mongoose.model('Sprint', schema);
